solvation_analysis/analysis.py

```diff
@@ -0,0 +1,321 @@
+            ax.set_title(f"Solvation distance of {name}")
+            self.rdf_plots[name] = fig, ax
+        assert self.solvents.keys() == self.radii.keys(), "Radii missing."
+
+    def _single_frame(self):
+        """
+        This function finds the solvation shells of each solute at a given time step.
+        """
+        # initialize empty arrays
+        all_pairs_list = []
+        all_dist_list = []
+        all_tags_list = []
+        # loop to find solvated atoms of each type
+        for name, solvent in self.solvents.items():
+            pairs, dist = capped_distance(
+                self.solute.positions,
+                solvent.positions,
+                self.radii[name],
+                box=self.u.dimensions,
+            )
+            # replace local ids with absolute ids
+            pairs[:, 1] = solvent.ids[[pairs[:, 1]]]  # TODO: ids vs ix?
+            # extend
+            all_pairs_list.append(pairs)
+            all_dist_list.append(dist)
+            all_tags_list.append(np.full(len(dist), name))  # creating a name array
+        all_pairs = np.concatenate(all_pairs_list, dtype=int)
+        all_dist = np.concatenate(all_dist_list)
+        all_tags = np.concatenate(all_tags_list)
+        frame_length = len(all_pairs)
+        all_frames = np.full(frame_length, self._ts.frame)
+        # put the data into a data frame
+        all_resid = self.u.atoms[all_pairs[:, 1]].resids
+        solvation_data_np = np.column_stack(
+            (all_frames, all_pairs[:, 0], all_pairs[:, 1], all_dist, all_tags, all_resid)
+        )
+        self.solvation_frames.append(solvation_data_np)
+
+    def _conclude(self):
+        """
+        Instantiates the SolvationData class and several analysis classes.
+        """
+        solvation_data_np = np.vstack(self.solvation_frames)
+        solvation_data_df = pd.DataFrame(
+            solvation_data_np,
+            columns=["frame", "solvated_atom", "atom_id", "dist", "res_name", "res_id"]
+        )
+        # clean up solvation_data df
+        for column in ["frame", "solvated_atom", "atom_id", "dist", "res_id"]:
+            solvation_data_df[column] = pd.to_numeric(solvation_data_df[column])
+        solvation_data_dup = solvation_data_df.sort_values(["frame", "solvated_atom", "dist"])
+        solvation_data = solvation_data_dup.drop_duplicates(["frame", "solvated_atom", "res_id"])
+        self.solvation_data_dup = solvation_data_dup.set_index(["frame", "solvated_atom", "atom_id"])
+        self.solvation_data = solvation_data.set_index(["frame", "solvated_atom", "atom_id"])
+        # create analysis classes
+        self.ion_speciation = _Speciation(self.solvation_data, self.n_frames, self.n_solute)
+        self.ion_pairing = _Pairing(self.solvation_data, self.n_frames, self.n_solute)
+        self.coordination_numbers = _Coordination(self.solvation_data, self.n_frames, self.n_solute)
+
+    def map_step_to_index(self, traj_step):
```

Comment:
I could be missing some thing but I don't really see the point of this function. If you want to analyse a frame you should specify the **exact** frame index. 

This has the potential to mislead people IMO where they think they are analysing a frame but they are analysing a neighbouring one.

Reply:
I understand that this is meant to solve the:
 ```python
Solution.selection_function(atom_index=1, step=700)
Solution.selection_function(atom_index=1, step=3) 
```
problem, but I think if you instead force **one** convention or allow the use of something like:

 ```python
Solution.selection_function(atom_index=1, step=700)
Solution.selection_function(atom_index=1, step=3, internal_index=True) 
```
It will be much clearer for the user. Let me know if anyone disagrees. At the very least I think any mapping should be out of the users reach.

Reply:
This function is now meant to map a frame of interest to a frame that was actually analyzed. If I used a step of 100 then I might be interested in step 753 but only step 700 is available. This gives me a convenient way to map onto analyzed steps.

With that said, I agree that it should be out of users reach.